File: project-java/src/main/java/com/chimerapps/alexa/home/SmartHomeRequestHandler.java
/*
 *    Copyright 2017 Chimerapps
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package com.chimerapps.alexa.home;

import com.amazonaws.services.lambda.runtime.Context;
import com.chimerapps.alexa.home.error.DriverInternalError;
import com.chimerapps.alexa.home.error.SmartHomeError;
import com.chimerapps.alexa.home.error.UnsupportedOperationError;
import com.chimerapps.alexa.home.model.ApplianceRequest;
import com.chimerapps.alexa.home.model.ResponsePayload;
import com.chimerapps.alexa.home.model.SmartHomeHeader;
import com.chimerapps.alexa.home.model.SmartHomeReply;
import com.chimerapps.alexa.home.model.SmartHomeRequest;
import com.chimerapps.alexa.home.model.discovery.DiscoverAppliancesRequest;
import com.chimerapps.alexa.home.model.discovery.DiscoverAppliancesResponse;
import com.chimerapps.alexa.home.model.healthcheck.HealthCheckRequest;
import com.chimerapps.alexa.home.model.healthcheck.HealthCheckResponse;
import com.chimerapps.alexa.home.model.lockstate.GetLockStateResponse;
import com.chimerapps.alexa.home.model.lockstate.SetLockStateConfirmation;
import com.chimerapps.alexa.home.model.lockstate.SetLockStateRequest;
import com.chimerapps.alexa.home.model.percentage.DecrementPercentageConfirmation;
import com.chimerapps.alexa.home.model.percentage.DecrementPercentageRequest;
import com.chimerapps.alexa.home.model.percentage.IncrementPercentageConfirmation;
import com.chimerapps.alexa.home.model.percentage.IncrementPercentageRequest;
import com.chimerapps.alexa.home.model.percentage.SetPercentageConfirmation;
import com.chimerapps.alexa.home.model.percentage.SetPercentageRequest;
import com.chimerapps.alexa.home.model.temperature.DecrementTargetTemperatureConfirmation;
import com.chimerapps.alexa.home.model.temperature.DecrementTargetTemperatureRequest;
import com.chimerapps.alexa.home.model.temperature.GetTargetTemperatureResponse;
import com.chimerapps.alexa.home.model.temperature.GetTemperatureReadingResponse;
import com.chimerapps.alexa.home.model.temperature.IncrementTargetTemperatureRequest;
import com.chimerapps.alexa.home.model.temperature.SetTargetTemperatureConfirmation;
import com.chimerapps.alexa.home.model.temperature.SetTargetTemperatureRequest;
import com.chimerapps.alexa.home.model.toggle.TurnOffConfirmation;
import com.chimerapps.alexa.home.model.toggle.TurnOnConfirmation;
import com.fasterxml.jackson.core.JsonProcessingException;

/**
 * @author Nicola Verbeeck
 * Date 10/03/2017.
 */
@SuppressWarnings({"WeakerAccess", "unused"})
public class SmartHomeRequestHandler extends RawSmartHomeRequestHandler {

	public static final String ACTION_GET_LOCK_STATE = "GetLockStateRequest";
	public static final String ACTION_SET_LOCK_STATE = "SetLockStateRequest";

	public static final String ACTION_GET_TARGET_TEMPERATURE = "GetTargetTemperatureRequest";
	public static final String ACTION_SET_TARGET_TEMPERATURE = "SetTargetTemperatureRequest";
	public static final String ACTION_INC_TEMPERATURE = "IncrementTargetTemperatureRequest";
	public static final String ACTION_DEC_TEMPERATURE = "DecrementTargetTemperatureRequest";

	public static final String ACTION_GET_TEMPERATURE_READING = "GetTemperatureReadingRequest";

	public static final String ACTION_HEALTH_CHECK = "HealthCheckRequest";

	public static final String ACTION_DEC_PERCENTAGE = "DecrementPercentageRequest";
	public static final String ACTION_INC_PERCENTAGE = "IncrementPercentageRequest";
	public static final String ACTION_SET_PERCENTAGE = "SetPercentageRequest";

	public static final String ACTION_TURN_ON = "TurnOnRequest";
	public static final String ACTION_TURN_OFF = "TurnOffRequest";


	protected DiscoverAppliancesResponse handleDiscovery(final SmartHomeHeader header, final DiscoverAppliancesRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected GetLockStateResponse handleGetLockState(final SmartHomeHeader header, final ApplianceRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected GetTargetTemperatureResponse handleGetTargetTemperature(final SmartHomeHeader header, final ApplianceRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected GetTemperatureReadingResponse handleGetTemperatureReading(final SmartHomeHeader header, final ApplianceRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected HealthCheckResponse handleHealthCheck(final SmartHomeHeader header, final HealthCheckRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected SetLockStateConfirmation handleSetLockState(final SmartHomeHeader header, final SetLockStateRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected IncrementPercentageConfirmation handleIncrementTemperature(final SmartHomeHeader header, final IncrementTargetTemperatureRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected DecrementTargetTemperatureConfirmation handleDecrementTemperature(final SmartHomeHeader header, final DecrementTargetTemperatureRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected DecrementPercentageConfirmation handleDecrementPercentage(final SmartHomeHeader header, final DecrementPercentageRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected IncrementPercentageConfirmation handleIncrementPercentage(final SmartHomeHeader header, final IncrementPercentageRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected SetPercentageConfirmation handleSetPercentage(final SmartHomeHeader header, final SetPercentageRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected TurnOnConfirmation handleTurnOn(final SmartHomeHeader header, final ApplianceRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected TurnOffConfirmation handleTurnOff(final SmartHomeHeader header, final ApplianceRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}

	protected SetTargetTemperatureConfirmation handleSetTemperature(final SmartHomeHeader header, final SetTargetTemperatureRequest request, final Context context) throws SmartHomeError {
		throw new UnsupportedOperationError(header, "Operation not supported");
	}


	@Override
	protected SmartHomeReply onDiscovery(final SmartHomeRequest request, final Context context) throws SmartHomeError {
		try {
			final DiscoverAppliancesRequest actualRequest = mapper.treeToValue(request.getPayload(), DiscoverAppliancesRequest.class);
			return new SmartHomeReply(request.getHeader(), handleDiscovery(request.getHeader(), actualRequest, context));
		} catch (final JsonProcessingException e) {
			throw new DriverInternalError(request.getHeader(), "Failed to read json", e);
		}
	}

	@Override
	protected SmartHomeReply onQuery(final SmartHomeRequest request, final Context context) throws SmartHomeError {
		try {
			final ResponsePayload payload;
			switch (request.getHeader().getName()) {
				case ACTION_GET_LOCK_STATE:
					payload = handleGetLockState(request.getHeader(), mapper.treeToValue(request.getPayload(), ApplianceRequest.class), context);
					break;
				case ACTION_GET_TARGET_TEMPERATURE:
					payload = handleGetTargetTemperature(request.getHeader(), mapper.treeToValue(request.getPayload(), ApplianceRequest.class), context);
					break;
				case ACTION_GET_TEMPERATURE_READING:
					payload = handleGetTemperatureReading(request.getHeader(), mapper.treeToValue(request.getPayload(), ApplianceRequest.class), context);
					break;
				default:
					throw new UnsupportedOperationError(request.getHeader(), "Unknown name: " + request.getHeader().getName());
			}
			return new SmartHomeReply(request.getHeader(), payload);
		} catch (final JsonProcessingException e) {
			throw new DriverInternalError(request.getHeader(), "Failed to read json", e);
		}
	}

	@Override
	protected SmartHomeReply onControl(SmartHomeRequest request, Context context) throws SmartHomeError {
		try {
			final ResponsePayload payload;
			switch (request.getHeader().getName()) {
				case ACTION_SET_LOCK_STATE:
					payload = handleSetLockState(request.getHeader(), mapper.treeToValue(request.getPayload(), SetLockStateRequest.class), context);
					break;
				case ACTION_SET_TARGET_TEMPERATURE:
					payload = handleSetTemperature(request.getHeader(), mapper.treeToValue(request.getPayload(), SetTargetTemperatureRequest.class), context);
					break;
				case ACTION_INC_TEMPERATURE:
					payload = handleIncrementTemperature(request.getHeader(), mapper.treeToValue(request.getPayload(), IncrementTargetTemperatureRequest.class), context);
					break;
				case ACTION_DEC_TEMPERATURE:
					payload = handleDecrementTemperature(request.getHeader(), mapper.treeToValue(request.getPayload(), DecrementTargetTemperatureRequest.class), context);
					break;
				case ACTION_DEC_PERCENTAGE:
					payload = handleDecrementPercentage(request.getHeader(), mapper.treeToValue(request.getPayload(), DecrementPercentageRequest.class), context);
					break;
				case ACTION_INC_PERCENTAGE:
					payload = handleIncrementPercentage(request.getHeader(), mapper.treeToValue(request.getPayload(), IncrementPercentageRequest.class), context);
					break;
				case ACTION_SET_PERCENTAGE:
					payload = handleSetPercentage(request.getHeader(), mapper.treeToValue(request.getPayload(), SetPercentageRequest.class), context);
					break;
				case ACTION_TURN_ON:
					payload = handleTurnOn(request.getHeader(), mapper.treeToValue(request.getPayload(), ApplianceRequest.class), context);
					break;
				case ACTION_TURN_OFF:
					payload = handleTurnOff(request.getHeader(), mapper.treeToValue(request.getPayload(), ApplianceRequest.class), context);
					break;
				default:
					throw new UnsupportedOperationError(request.getHeader(), "Unknown name: " + request.getHeader().getName());
			}
			return new SmartHomeReply(request.getHeader(), payload);
		} catch (final JsonProcessingException e) {
			throw new DriverInternalError(request.getHeader(), "Failed to read json", e);
		}
	}

	@Override
	protected SmartHomeReply onSystem(SmartHomeRequest request, Context context) throws SmartHomeError {
		try {
			final ResponsePayload payload;
			switch (request.getHeader().getName()) {
				case ACTION_HEALTH_CHECK:
					payload = handleHealthCheck(request.getHeader(), mapper.treeToValue(request.getPayload(), HealthCheckRequest.class), context);
					break;
				default:
					throw new UnsupportedOperationError(request.getHeader(), "Unknown name: " + request.getHeader().getName());
			}
			return new SmartHomeReply(request.getHeader(), payload);
		} catch (final JsonProcessingException e) {
			throw new DriverInternalError(request.getHeader(), "Failed to read json", e);
		}
	}
}

File: project-kotlin/src/main/kotlin/com/chimerapps/easypus/alexa/SmartHomeRequestHandler.kt
/*
 *    Copyright 2017 Chimerapps
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package com.chimerapps.easypus.alexa

import com.amazonaws.services.lambda.runtime.Context
import com.chimerapps.easypus.alexa.error.SmartHomeError
import com.chimerapps.easypus.alexa.error.UnsupportedOperationError
import com.chimerapps.easypus.alexa.model.*
import com.chimerapps.easypus.alexa.utils.readValue

/**
 * @author Nicola Verbeeck
 * @date 09/03/2017.
 */
@Suppress("unused", "UNUSED_PARAMETER")
abstract class SmartHomeRequestHandler : RawSmartHomeRequestHandler() {

    @Throws(SmartHomeError::class)
    protected fun handleDiscovery(header: SmartHomeHeader, request: DiscoverAppliancesRequest, context: Context): DiscoverAppliancesResponse = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleGetLockState(header: SmartHomeHeader, request: GetLockStateRequest, context: Context): GetLockStateResponse = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleGetTargetTemperature(header: SmartHomeHeader, request: GetTargetTemperatureRequest, context: Context): GetTargetTemperatureResponse = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleGetTemperatureReading(header: SmartHomeHeader, request: GetTemperatureReadingRequest, context: Context): GetTemperatureReadingResponse = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleHealthCheck(header: SmartHomeHeader, request: HealthCheckRequest, context: Context): HealthCheckResponse = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleSetLockState(header: SmartHomeHeader, request: SetLockStateRequest, context: Context): SetLockStateConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleIncrementTemperature(header: SmartHomeHeader, request: IncrementTargetTemperatureRequest, context: Context): IncrementPercentageConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleDecrementTemperature(header: SmartHomeHeader, request: DecrementTargetTemperatureRequest, context: Context): DecrementTargetTemperatureConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleDecrementPercentage(header: SmartHomeHeader, request: DecrementPercentageRequest, context: Context): DecrementPercentageConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleIncrementPercentage(header: SmartHomeHeader, request: IncrementPercentageRequest, context: Context): IncrementPercentageConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleSetPercentage(header: SmartHomeHeader, request: SetPercentageRequest, context: Context): SetPercentageConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleTurnOn(header: SmartHomeHeader, request: TurnOnRequest, context: Context): TurnOnConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleTurnOff(header: SmartHomeHeader, request: TurnOffRequest, context: Context): TurnOffConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    protected fun handleSetTemperature(header: SmartHomeHeader, request: SetTargetTemperatureRequest, context: Context): SetTargetTemperatureConfirmation = throw UnsupportedOperationError(header, "Operation not supported")

    @Throws(SmartHomeError::class)
    override fun onDiscovery(request: SmartHomeRequest, context: Context): SmartHomeReply {
        val actualRequest = mapper.readValue<DiscoverAppliancesRequest>(request.payload)
        return SmartHomeReply(request.header, handleDiscovery(request.header, actualRequest, context))
    }

    @Throws(SmartHomeError::class)
    override fun onQuery(request: SmartHomeRequest, context: Context): SmartHomeReply {
        return SmartHomeReply(request.header,
                when (request.header.name) {
                    ACTION_GET_LOCK_STATE -> handleGetLockState(request.header, mapper.readValue<GetLockStateRequest>(request.payload), context)
                    ACTION_GET_TARGET_TEMPERATURE -> handleGetTargetTemperature(request.header, mapper.readValue<GetTargetTemperatureRequest>(request.payload), context)
                    ACTION_GET_TEMPERATURE_READING -> handleGetTemperatureReading(request.header, mapper.readValue<GetTemperatureReadingRequest>(request.payload), context)
                    else -> throw UnsupportedOperationError(request.header, "Unknown name: ${request.header.name}")
                })
    }

    @Throws(SmartHomeError::class)
    override fun onControl(request: SmartHomeRequest, context: Context): SmartHomeReply {
        return SmartHomeReply(request.header,
                when (request.header.name) {
                    ACTION_SET_LOCK_STATE -> handleSetLockState(request.header, mapper.readValue<SetLockStateRequest>(request.payload), context)
                    ACTION_SET_TARGET_TEMPERATURE -> handleSetTemperature(request.header, mapper.readValue<SetTargetTemperatureRequest>(request.payload), context)
                    ACTION_INC_TEMPERATURE -> handleIncrementTemperature(request.header, mapper.readValue<IncrementTargetTemperatureRequest>(request.payload), context)
                    ACTION_DEC_TEMPERATURE -> handleDecrementTemperature(request.header, mapper.readValue<DecrementTargetTemperatureRequest>(request.payload), context)
                    ACTION_DEC_PERCENTAGE -> handleDecrementPercentage(request.header, mapper.readValue<DecrementPercentageRequest>(request.payload), context)
                    ACTION_INC_PERCENTAGE -> handleIncrementPercentage(request.header, mapper.readValue<IncrementPercentageRequest>(request.payload), context)
                    ACTION_SET_PERCENTAGE -> handleSetPercentage(request.header, mapper.readValue<SetPercentageRequest>(request.payload), context)
                    ACTION_TURN_ON -> handleTurnOn(request.header, mapper.readValue<TurnOnRequest>(request.payload), context)
                    ACTION_TURN_OFF -> handleTurnOff(request.header, mapper.readValue<TurnOffRequest>(request.payload), context)
                    else -> throw UnsupportedOperationError(request.header, "Unknown name: ${request.header.name}")
                }
        )
    }

    @Throws(SmartHomeError::class)
    override fun onSystem(request: SmartHomeRequest, context: Context): SmartHomeReply {
        return SmartHomeReply(request.header,
                when (request.header.name) {
                    ACTION_HEALTH_CHECK -> handleHealthCheck(request.header, mapper.readValue<HealthCheckRequest>(request.payload), context)
                    else -> throw UnsupportedOperationError(request.header, "Unknown name: ${request.header.name}")
                })
    }

    companion object {
        val ACTION_GET_LOCK_STATE = "GetLockStateRequest"
        val ACTION_SET_LOCK_STATE = "SetLockStateRequest"

        val ACTION_GET_TARGET_TEMPERATURE = "GetTargetTemperatureRequest"
        val ACTION_SET_TARGET_TEMPERATURE = "SetTargetTemperatureRequest"
        val ACTION_INC_TEMPERATURE = "IncrementTargetTemperatureRequest"
        val ACTION_DEC_TEMPERATURE = "DecrementTargetTemperatureRequest"

        val ACTION_GET_TEMPERATURE_READING = "GetTemperatureReadingRequest"

        val ACTION_HEALTH_CHECK = "HealthCheckRequest"

        val ACTION_DEC_PERCENTAGE = "DecrementPercentageRequest"
        val ACTION_INC_PERCENTAGE = "IncrementPercentageRequest"
        val ACTION_SET_PERCENTAGE = "SetPercentageRequest"

        val ACTION_TURN_ON = "TurnOnRequest"
        val ACTION_TURN_OFF = "TurnOffRequest"
    }
}